repository/src/fetch.rs

```diff
@@ -0,0 +1,269 @@
+        .map(|(commit, hash)| {
+            from_semantic_commit(commit)
+                .map_err(|e| (e, hash))
+                .map(|x| (x, hash))
+        })
+        .collect::<Result<Vec<_>, _>>()
+        .map_err(|(e, c)| CommitError::Commit(e, c))?;
+    Ok(commits)
+}
+
+pub async fn fetch<T: RawRepository>(
+    this: &mut DistributedRepository<T>,
+    _network_config: &NetworkConfig,
+    known_peers: &[Peer],
+) -> Result<(), Error> {
+    add_remotes(this, known_peers).await?;
+    this.raw.fetch_all().await?;
+
+    let remote_branches = this.raw.list_remote_tracking_branches().await?;
+    let finalization_proofs = retrieve_fps(this, &remote_branches).await?;
+    // Retrieve local branches to skip if the remote tracking branch is already in the local repo.
+    let local_branches = retrieve_local_branches(&this.raw).await?;
+
+    let mut last_finalized_commit_hash =
+        this.raw.locate_branch(FINALIZED_BRANCH_NAME.into()).await?;
+    let mut last_finalization_proof = format::fp_from_semantic_commit(
+        this.raw
+            .read_semantic_commit(this.raw.locate_branch(FP_BRANCH_NAME.into()).await?)
+            .await?,
+    )?;
+    let last_header = this.get_last_finalized_block_header().await?;
+    let reserved_state = this.get_reserved_state().await?;
+
+    let mut next_step_branches = Vec::new();
+
+    // step 1: update finalization
+    'branch_loop: for (remote_name, branch_name, commit_hash) in remote_branches {
+        let branch_displayed = format!(" {}/{}(at {})", remote_name, branch_name, commit_hash);
+
+        // Skip if the branch is `fp`.
+        if branch_name == FP_BRANCH_NAME {
+            continue;
+        }
+
+        // Check if the branch is already in the local repo.
+        if local_branches
+            .iter()
+            .any(|(_, commit_hash_)| *commit_hash_ == commit_hash)
+        {
+            info!("skip {}: alreadd tracked by local repo", branch_displayed);
+            continue;
+        }
+
+        // Check if the branch is rebased on the finalized branch.
+        if last_finalized_commit_hash
+            != this
+                .raw
+                .find_merge_base(last_finalized_commit_hash, commit_hash)
+                .await?
+        {
+            info!("remote tracking branch outdated: {}", branch_displayed);
+            continue;
+        }
+
+        // Reads the commits in the branch.
+        let commits = match read_commits(this, last_finalized_commit_hash, commit_hash).await {
+            Ok(x) => x,
+            Err(CommitError::Commit(error, commit)) => {
+                warn!("failed to parse commit {}: {}", commit, error);
+                continue;
+            }
+            Err(e) => return Err(e.into()),
+        };
+
+        // Verify all the incoming commits
+        let mut csv = CommitSequenceVerifier::new(last_header.clone(), reserved_state.clone())
+            .expect("finalized branch is not accepted by CSV");
```

Comment:
```suggestion
            .expect("finalized branch is not accepted by Commit Sequence Verifier");
```

Reply:
`.expect()` message is NOT for the user but for the developers.

Reply:
oh then how is the error handled?

Reply:
`.expect()` is not to be 'handled'. It is used if we encounter an error that can't be handled.
More directly saying, if `expect()` hits, that is an obvious bug.

Reply:
understood